// jupyter-iframe-commands/lib/index.js
// Copyright (c) TileDB, Inc.
// Distributed under the terms of the Modified BSD License.
import { ISettingRegistry } from '@jupyterlab/settingregistry';
import { expose, windowEndpoint } from 'comlink';
/**
 * A plugin to expose an API for interacting with JupyterLab from a parent page.
 */
const plugin = {
    id: 'jupyter-iframe-commands:plugin',
    autoStart: true,
    description: 'A plugin to expose an API for interacting with JupyterLab from a parent page.',
    optional: [ISettingRegistry],
    activate: async (app, settingRegistry) => {
        console.log('JupyterLab extension jupyter-iframe-commands is activated!');
        const { commands } = app;
        if (settingRegistry) {
            settingRegistry
                .load(plugin.id)
                .then(settings => {
                console.log('jupyter-iframe-commands settings loaded:', settings.composite);
            })
                .catch(reason => {
                console.error('Failed to load settings for jupyter-iframe-commands.', reason);
            });
        }
        const api = {
            async execute(command, args) {
                const result = await commands.execute(command, args);
                try {
                    // Attempt to clone result to check if it's serializable
                    const clone = structuredClone(result);
                    return clone;
                }
                catch (error) {
                    // Non-serializable values can't be sent over Comlink
                    void error;
                }
            },
            async listCommands() {
                return commands.listCommands();
            },
            get ready() {
                return app.started;
            }
        };
        app.started.then(() => {
            var _a;
            const endpoint = windowEndpoint(self.parent);
            expose(api, endpoint);
            (_a = window.parent) === null || _a === void 0 ? void 0 : _a.postMessage('_JUPYTER_IFRAME_COMMANDS_LOADED', '*');
        });
    }
};
export default plugin;
export * from './interface';
